### DISCUSSION TITLE
Adding `KModifier` for interface type parameter.

### DISCUSSION THREAD

**Discussion by vnermolaev** [square/kotlinpoet#1208] [Q&A] [answered]
Hi!

Somewhere in the code, I have interfaces:
 
```kotlin
interface Base

interface BaseProvider {
    fun list(): List<out Base>
}
```

now using kotlin poet, I'd like to create an implementation of `BaseProvider`, but I am facing difficulties trying to specify the return type of `list()`, consider the following fragment:
```kotlin
List::class.parameterizedBy(
    Base::class  // <- how do I add a modifier KModifier.OUT 
)
```

One thing I tried is to "abuse" `TypeVariableName` as follows
```kotlin
List::class.asClassName().parameterizedBy(
    TypeVariableName(T::class.qualifiedName.toString(), KModifier.OUT)
)
```
but in this case, the modifier is stripped and the results of the two snippets are equivalent.

Is there a mechanism or a workaround to variance in such cases?

**Comment by Egorand** [comment, marked as answer]
What you're looking for is probably something along the lines of:

```kotlin
List::class.asClassName().parameterizedBy(
    WildcardTypeName.producerOf(Base::class.asTypeName())
)
```

**Reply by vnermolaev** [reply]
Thanks a lot! Unfortunately, this is not documented :(

**Reply by Egorand** [reply]
There is a mention of `WildcardTypeName.producerOf` in [%T for Types](https://square.github.io/kotlinpoet/#t-for-types), but please feel free to send us a PR if you have ideas on how to document it better.
